third_party/ibis/ibis_impala/api.py

```diff
@@ -105,6 +107,81 @@ def _fetch(self, cursor):
         df = _column_batches_to_dataframe(names, batches)
         return df
 
+def _column_batches_to_dataframe(names, batches):
+    cols = {}
+    for name, chunks in zip(names, zip(*[b.columns for b in batches])):
+        cols[name] = _chunks_to_pandas_array(chunks)
+    return pd.DataFrame(cols, columns=names)
+
+
+def _chunks_to_pandas_array(chunks):
+    total_length = 0
+    have_nulls = False
+    for c in chunks:
+        total_length += len(c)
+        have_nulls = have_nulls or c.nulls.any()
+
+    type_ = chunks[0].data_type
+    numpy_type = _HS2_TTypeId_to_dtype[type_]
+
+    def fill_nonnull(target, chunks):
```

Comment:
Everything that is unchanged should be imported rather than re-written IMO, since we plan to push this upstream asap

Reply:
+1, but in this case it's a nested function so I'm not aware of a way to import that

Reply:
Would be good if `_HS2_TTypeId_to_dtype` and `_column_batches_to_dataframe()` are imported then, but yes I'm also not sure what we can do about this nested function specifically.